### DISCUSSION TITLE
Suggestion: Send reset email to multiple addresses / Recovery email address

### DISCUSSION THREAD

**Discussion by KoltesDigital** [authelia/authelia#5198] [Ideas]
Hi,

I've installed Authelia with success, among others as a light password self-service over LDAP (I can't ask users to change their password with phpLDAPadmin 😄). The problem is, I want to use that LDAP for the mail servers. If users forget their password, resetting it will send an email, but they can't access their mailbox, and they won't be able to finish the reset procedure.

I imagine two solutions:

1. as the `mail` attribute is multivalued, the reset notifications could be sent to every email address. I think the other notifications could keep on using the first one only, there's no need to spam users who still have access to their primary mailbox.
2. a configuration field (e.g. `authentication_backend.ldap.recovery_mail_attribute`) could define the attribute to use to send reset email. It seems there's no consensus regarding which field to use, so sysadmins could either add a new attribute in their schema, or be lazy and use a random unused attribute from inetOrgPerson. The configuration field allow them to choose freely.

I prefer the first approach, it looks less hacky regarding LDAP, et allow users to specify as many addresses as they want to, which may sometimes make sense. Any opinion?

**Comment by ivanov17** [comment]
I want to use Authelia for the same purposes and this behavior confused me a lot. I think if a user entry has multiple mail attributes, Authelia MUST send emails to each address. Authelia doesn't have to care about the number of user mailboxes. 

Any service that allows you to specify multiple contacts sends emails to each address. Google, Yahoo, GMX... I don't know exceptions. Why does Authelia only send email to the first address on the list? Why not twentieth, or third, or whatever? I think this is a bug.

Many thanks to the developers of Authelia for a great product, but this behavior is really weird.

**Comment by james-d-elliott** [comment]
The behavior is intentional as the email should only be sent to a single location as sending it to multiple has security implications. This issue was not closed as there is merit in the idea of allowing the *__user__* to manually verify additional email addresses specifically for both their identity profile and for emails.

This is not a bug.

Also as this is a suggestion it's being moved to ideas.

**Comment by ivanov17** [comment]
@james-d-elliott I think if the admin specifies some attribute as mail attribute and allows the user to have multiple addresses, it's not a security issue, it's expected behavior.

If someone can to add mail attribute to user entry, in most cases they can also change value of password attribute in the same entry. Authelia can't prevent this.

But if the user has an additional address and needs to reset password for primary address, they won't receive email at all, as Authelia considers this a security issue.

In this case, the intentional behavior looks like a bug.

**Reply by james-d-elliott** [reply]
A single action should only ever generate a single email. These emails contain information that can authorize the user to do something. This means if an attacker gains access to the emails, they have a portion of the authorization they require to attack a target. Every additional email is an increase to this attack surface.

Ergo, sending multiple emails for this purpose *__absolutely is__* a security issue. Any person, project, or company who does not recognize this is and instead blatantly ignoring some very basic security principles or is just ignorant of them.

**Reply by ivanov17** [reply]
@james-d-elliott I apologize if my comments sound harsh. English is not my native language. I didn't mean to offend anyone.

Comment hidden by you is about security and convenience. You know that these are mutually contradictory things. Security is not my specialty, I may to not understand something. But I see that sending multiple copies of a message containing sensitive information is a common practice. I use multiple addresses for my Github account and Github always send me a copy of any message for each one.

Mail is a multi-valued LDAP attribute. I expect an application that works with it to work with each of its values. There is logic in this, right?

If an application doesn't work properly with a multi-valued attribute, I think it's a bug. I don't expect that a multi-valued attribute to be treated as single-valued. Especially since there is nothing about it in the documentation.

You wrote that every additional email increases an attack surface. Yes, but any convenient thing reduces the level of security. It's safer to have a separate password for each online account and not use single sign-on at all.

I believe that administrators, not developers, should be responsible for such things. I should have a choice of security level in my infrastructure. So I think there should be a configuration option to let me choose.

What do you think about it?

**Reply by james-d-elliott** [reply]
> Comment hidden by you is about security and convenience. You know that these are mutually contradictory things. 

This is absolutely factually false. A feature can improve both of these things at the same time, as will be discussed later.

Authelia opts for a security by design mantra (and has always done this) as per the following:

- OWASP Best Practices
- OWASP Security by Design Principles
- Open Source Security Foundation Best Practices

We aim for a convenient security solution but prefer to avoid both convenient features which reduce security, and opt in features that are hard to properly test. This idea fits into **_both_** of these categories.

> Security is not my specialty, I may to not understand something. But I see that sending multiple copies of a message containing sensitive information is a common practice. I use multiple addresses for my Github account and Github always send me a copy of any message for each one.

What's common is not evidence of good practices. A bad practice is also a bad practice regardless of convenience. There is such thing as a common bad practice. As far as I am aware none of these services allow you to send information that could be used to escalate an attackers privileges to multiple addresses, they only send informational messages.

> Mail is a multi-valued LDAP attribute. I expect an application that works with it to work with each of its values. There is logic in this, right?

Sure there is logic, but that does not counter there being logic in it being inherently a bad security practice. Just because a point has logic does not mean it's naturally a counter to all other logic.

> If an application doesn't work properly with a multi-valued attribute, I think it's a bug. I don't expect that a multi-valued attribute to be treated as single-valued. Especially since there is nothing about it in the documentation.

A bug is when documented, standardized, or intended behavior does not work as expected. This doesn't apply to this. It's not documented to behave in this way, it's not a ratified standard, and it's not intended behavior. Just because you want something to behave in a certain way does not make it a bug. 

> You wrote that every additional email increases an attack surface. Yes, but any convenient thing reduces the level of security. It's safer to have a separate password for each online account and not use single sign-on at all.

If you apply the `convenience` and  `a common practice` arguments to passwords, that would suggest it's reasonable to trade off the security of hashed passwords (i.e. instead store them in plain text) for the benefit of being able to recover them. I don't believe the convenience argument is a particularly compelling one at all to ignore a bad practice. Neither do I think that if GitHub/Google etc did it we should automatically follow their lead is a compelling argument (i.e. "common practice"). 

> I believe that administrators, not developers, should be responsible for such things. I should have a choice of security level in my infrastructure. So I think there should be a configuration option to let me choose.
> 
> What do you think about it?

If we implement it we'll likely add a hidden flag which enables it, which will be littered with warnings about how it's a bad practice. We will also likely handle it in a way where the end-user (not the admin) must have given informed consent to this bad practice. This alleviates most of the security issues, at least it does completely from a compliance perspective.

As to priorities, making it send to multiple emails is basically not even on the list of things we're going to do. Only 4 people are interested and it is a complex feature to implement securely. 

It's more likely as far more people are interested in it, that we'll allow end-users to manually override their security specific email address, as this actually has a security and convenience benefit.

**Reply by ivanov17** [reply]
@james-d-elliott Thank you for detailed and intelligible explanation. 

But I have some comments and questions.

> As far as I am aware none of these services allow you to send information that could be used to escalate an attackers privileges to multiple addresses, they only send informational messages.

Yes, they usually ask for additional verification when it comes to password recovery or something like that. I rarely need these things, but they are useful in case of an emergency (disk crash or something else). But any major service sends, for example, messages with verification codes to all my registered addresses.

> Sure there is logic, but that does not counter there being logic in it being inherently a bad security practice.

Okay, I agree that this is not the best practice. But there is the first problem: multi-valued attribute is processed as single-valued without any warnings. Just imagine that for some reasons a multi-valued attribute used with multiple values. Authelia takes the first value and do nothing to all the others. This is also not the best practice.

I think I should somehow warn my users that they will only recieve messages to first mailbox in the alphabetical order. Also, I see another problem here. Users cannot choose which address is the primary one.

Okay, single action means single message. But what I should do if user has multiple email addresses? I think they should at least get an informational message that the link was sent to some address.

> A bug is when documented, standardized, or intended behavior does not work as expected.

Sorry, I didn't find anywhere in the documentation that Authelia only uses first value of multi-valued attribute. I always guess that multi-valued attribute processed as multi-valued. I think it's complies with the standard, and if I see different behavior I think that it's a bug.

I don't call you to do bad things that break security. It's just that I see a problem with using Authelia if the LDAP configuration is different from what you're suggesting. People use LDAP if they need in single authentication/authorization for multiple services. They can also use LDAP authentication for mail protocols (and multiple mail addresses), not just web accounts. If we could find a solution to make Authelia work with these usage patterns, that would be great.

**Reply by james-d-elliott** [reply]
1. If you provide a list of areas where you don't feel this specific issue is not well documented I will make an adjustment to clarify it with the provision that it seems appropriate.
2. We have already clearly explained what our short term intent is and that we've not decided on the long term yet. No sense in us explaining that again. The only thing I'll add is that all backends will have to support what ever we choose to do. You're welcome to subscribe to this issue and wait for updates, however you will not get an answer or solution in the short term.
3. All of the maintainers are very competent in quite a lot of technical matters including LDAP. I would say I'm probably the most competent when it comes to LDAP but that's hard to say or measure for certain. I also would not be surprised if I wasn't.
4. There are numerous ways to use LDAP. Multiple flavors, you can make your own, etc. Naturally it's unreasonable to expect any product free or paid will support every LDAP schema and usage. When you decide to use it in a way that is somewhat uncommon you can reasonably expect issues of your own creation.
5. Regarding what you can do to get this working in your situation, that's mostly up to you. Either make the first (we use sequential order as returned from your LDAP server, we do not sort alphabetically) email their primary email (this is a relatively common practice from what I've seen, but there are exceptions and it's a very subjective thing depending on your exposure), or make a single valued attribute their email for this purpose.

> I think they should at least get an informational message that the link was sent to some address.

They do get one.

> Sorry, I didn't find anywhere in the documentation that Authelia only uses first value of multi-valued attribute. I always guess that multi-valued attribute processed as multi-valued. I think it's complies with the standard, and if I see different behavior I think that it's a bug.

It **_is not_** documented that Authelia sends emails to multiple addresses is my point. If it's not documented to behave in a certain way and it is reasonable that someone else may not have configured it like you have, and the other criteria do not apply, regardless of project it's most likely not a bug.

**Reply by ivanov17** [reply]
> When you decide to use it in a way that is somewhat uncommon you can reasonably expect issues of your own creation.

Sorry, I use `mail` attribute  for its intended purpose: as multi-valued attribute. I use multiple values. What am I doing wrong?

I expect that the application to work with this attribute correctly, with each of its values. In the examples on your site, I see that the `mail` attribute is used for mail addresses, and not any other.

I thought that there was some kind of oversight and you didn't have time to fix it. But you tell me that I use LDAP in a way that is somewhat uncommon. I believe that using a multi-valued attribute as a single-valued is a way that is somewhat uncommon. 

Okay, we are four guys from all over the Internet who care about this. But this is a use case, so this case needs to be taken into account.

So, user entries have multiple values of mail attribute. What to do with it? This is just not handled. How is it going to be handled? This is not planned. Just use it or leave it. Okay, I understood. Thank you.

**Reply by james-d-elliott** [reply]
As per our [code of conduct](https://github.com/authelia/authelia/blob/master/CODE_OF_CONDUCT.md) we expect constructive feedback, professionalism, and respect of those interacting in our community.

You are being completely disrespectful, rude, condescending, and acting entitled, and this will be corrected; I hope that you choose to correct this and we are not forced to have to do this for you.

I've told you the solutions, I've asked for specific feedback, and I've answered your questions. I'm done with this discussion.

**Comment by ivanov17** [comment]
@james-d-elliott I think if the admin wants to limit the number of addresses allowed to receive emails, they should to do it explicitly. If it considered as a security issue, a configuration option is needed to allow/deny sending to multiple addresses. 

A service account that is allowed to read and write user attribute values can also be a security issue. Allowing a user to set a password from the web can be another security issue. Receiving emails with a password reset link to selected addresses is nothing special. The most secure server is a stopped server.

**Reply by james-d-elliott** [reply]
Keep it on topic, and please read our [code of conduct](https://github.com/authelia/authelia/blob/master/CODE_OF_CONDUCT.md).

**Comment by sphakka** [comment]
I second this idea of having at least an alternative mail channel for password reset tokens because, in practice, users forget their password and get locked out from their (main) mail account where the tokens are landing. Of course, any such extra channel should be properly verified, possibly, at enrollment time in the auth base -- I don't know if there's any trigger mechanism available for that in [Open]LDAP.  

Please, cf. Self Service Password which supports [configurable LDAP mail attributes](https://github.com/ltb-project/self-service-password/blob/master/conf/config.inc.php). I use it with an extended LDAP schema with a `mailAlternate` attribute: only this one is used for password reset tokens. 

IMHO, authelia, while acting as an _auth proxy_, should not be concerned with the implications of using a secondary mail channel. However, as an _identity provider_ it should have a mail (or any other channel) verification / enrollment mechanism.

**Reply by ivanov17** [reply]
> I don't know if there's any trigger mechanism available for that in [Open]LDAP

I think there is only one way to check anything here. We should be able to check the availability of an email address, but not on the LDAP server, it can't do that. So this should be a job for the identity provider. At a minimum, it should make a callback verification. A much better option is to send a message containing a verification code.

Another option is to trust any address that specified in the LDAP entry. Authelia only does this for one address. I really don't understand why any other address from the same entry is less trusted if the LDAP database is supposed to be managed by an administrator. Okay, this increases the attack surface. But any additional address increases it. As a result, at the moment any user can only recover their password if they personally come to the administrator.

Unfortunately, our conversation with James reached a dead end. As far as I understand, he does not consider this problem significant and claims that I am doing something strange. What to do, it seems that among lightweight identity providers, only Authelia uses LDAP as a database.

I think at the moment this issue can only be resolved by third-party solutions. For example, there is a package django-ldapdb for Django. Django can send notifications by email. All that remains is to write the necessary logic.

**Reply by james-d-elliott** [reply]
> But any additional address increases it. As a result, at the moment any user can only recover their password if they personally come to the administrator.

It's quite simple. Sending an email to 100 addresses means there could be up to 100 email addresses which have the details any of which being compromised would result in a security failure. We will only ever send it to 1 intentionally, not multiple. As such this feature will be closed as not being one we're going to implement and I will open as second discussion which will be along the lines of the feature I just described below (see #6093).

> Unfortunately, our conversation with James reached a dead end. As far as I understand, he does not consider this problem significant and claims that I am doing something strange. What to do, it seems that among lightweight identity providers, only Authelia uses LDAP as a database.

This is absolutely false. I very clearly stated the likely solution we would implement was that the user would have to register an address themselves (i.e. they add it and confirm it). This seems like a deliberate mischaracterization.
